rmw_connextdds_common/src/common/rmw_context.cpp

```diff
@@ -339,48 +361,9 @@ rmw_connextdds_initialize_participant_qos(
 }
 
 rmw_ret_t
-rmw_context_impl_t::initialize_node(
-  const rmw_discovery_options_t * const discovery_options_in)
+rmw_context_impl_s::initialize_node()
 {
-  if (this->node_count > 0) {
-    bool params_equal = false;
```

Comment:
This is the part that sticks out to me, these blocks seem to be just deleted, everything else in the diff seems to be just moved around/names changed.

I'm sure I just don't understand this code well enough, but could you explain why we no longer need this discovery options validation / initialization?

Reply:
Right, it's a good question, and is actually the heart of the fix here.

Prior to this PR, the code was actually doing two things with the discovery options that were wrong:
1.  It was allocating the discovery options when nodes were added.  But that really doesn't make any sense, since these discovery options are a per-context concept, not a per-node concept.  This is what led to the awkward code to check if discovery options are the same between nodes.  This is also what led to the memory leak, since in certain circumstances (adding one node, removing that node, and then adding another), we weren't properly freeing the associated memory.
2.  The discovery options were already stored in the `rmw_context_t` structure, in the `options.discovery_options` field, so there is no need to copy them at all here.  (I guess this isn't strictly "wrong", but is totally unnecessary)

To fix both of these, I changed it so that discovery options are always handled on a per-context basis.  That allows us to remove the code to check whether nodes in the same context have the same discovery options (since they do by definition), and also to get rid of the copy (since we just fetch it out of the `base_` member).

Does that make sense?